### PR TITLE
BentoML Deployment CLI POC

### PR DESCRIPTION
**Pull request by yubozhao**
(Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below.)

What changes were proposed in this pull request?
------------------------------------------------

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------

### CODE REVIEW

File: bentoml/cli/__init__.py [hunk 1]
```diff
@@ -168,10 +174,11 @@ def create_bentoml_cli():
     # Commands created here aren't mean to be used from generated service archive. They
     # are used as part of BentoML cli commands only.
 
-    add_deployment_commands(_cli)
+    add_legacy_deployment_commands(_cli)
     add_configuration_commands(_cli)
 
-    # pylint: enable=unused-variable
+    add_deployment_sub_commands(_cli)
+    add_configuration_sub_commands(_cli)
```
Comment (parano) [inline bentoml/cli/__init__.py]:
```
config_command = get_config_subcommand()

_cli.add_command(config_command)
```

Comment (yubozhao) [reply bentoml/cli/__init__.py]: done

File: bentoml/cli/deployment.py [hunk 2]
```diff
@@ -213,3 +227,211 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        print(bento_name, bento_verison)
+        operator_config = get_operator_config(platform, ctx.args)
+        spec = {
+            "bento_name": 'name',
+            "bento_version": 'version',
+            "operator": get_deployment_operator_type(platform),
+            "deployment_operator_config": operator_config,
+        }
+        result = get_yatai_service().ApplyDeployment(
+            request={
```
Comment (parano) [inline bentoml/cli/deployment.py]: ApplyDeploymentRequest(deployment=Deployment(....))

File: bentoml/cli/deployment.py [hunk 3]
```diff
@@ -213,3 +227,211 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        else:
+            _echo(
+                'Successful apply deployment {}'.format(bento_name), CLI_COLOR_SUCCESS
+            )
+            display_deployment_info(result.deployment)
+
+    @deployment.command()
+    @click.option("--name", type=click.STRING, help="Deployment's name", required=True)
+    def delete(name):
+        result = get_yatai_service().DeleteDeployment(request={'deployment_name': name})
```
Comment (parano) [inline bentoml/cli/deployment.py]: request = DeleteDeploymentRequest(...)
result = get_yatai_service().DeleteDeployment(request)

File: bentoml/cli/deployment.py [hunk 4]
```diff
@@ -213,3 +227,211 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        pass
+
+    @deployment.command(
+        help="Apply deployment configuration to target deployment platform",
+        short_help="Apply deployment configuration",
+        context_settings=dict(ignore_unknown_options=True, allow_extra_args=True),
+    )
+    @click.argument('bento-tag', type=click.STRING, required=True)
+    @click.option("--namespace", type=click.STRING, help="Deployment's namespace")
+    @click.option("--deployment-name", type=click.STRING, help="Deployment's name")
```
Comment (parano) [inline bentoml/cli/deployment.py]: => argument?

File: bentoml/cli/deployment.py [hunk 5]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+    return DeploymentOperator.Value(platform.upper())
+
+
+def display_response_status_error(status):
+    _echo('Error code: {}'.format(status.status_code), CLI_COLOR_ERROR)
+    _echo('Error message: {}'.format(status.error_message), CLI_COLOR_ERROR)
+
+
+def display_deployment_info(deployment):
+    _echo('Deployment {} info\n\n'.format(deployment.name))
```
Comment (parano) [inline bentoml/cli/deployment.py]: use YAML module instead? might be good to add a utility function to convert a Deployment_PB object to JSON or YAML

Comment (parano) [reply bentoml/cli/deployment.py]: we might need both to_json and from_json, I'm not sure if python protobuf provides that, could you look into it?

File: bentoml/cli/deployment.py [hunk 6]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        _echo('  State:')
+        _echo('    Current State: {}'.format(deployment.state.state))
+        if deployment.state.error_message:
+            _echo(
+                '    Error: {}'.format(deployment.state.error_message), CLI_COLOR_ERROR
+            )
+        _echo('    Info: {}'.format(deployment.state.info_json))
+
+
+def parse_bento_tag(tag):
```
Comment (parano) [inline bentoml/cli/deployment.py]: would be great to catch the error when there's multiple ':' in the tag?
```
def parse_bento_tag(tag):
       items = tag.split(':')
       if len(items) == 1:
           items.append('latest')
       if len(items) != 2:
           raise ValueError(...)
       return items
```

File: bentoml/cli/deployment.py [hunk 7]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+            DescribeDeploymentRequest(deployment_name=name)
+        )
+        if result.status.status_code != Status.OK:
+            _echo('Describe deployment {} failed'.format(name), CLI_COLOR_ERROR)
+            display_response_status_error(result.status)
+        else:
+            display_deployment_info(result.deployment)
+
+    @deploy.command()
+    @click.option("--limit", type=click.INT, help="")
```
Comment (parano) [inline bentoml/cli/deployment.py]: add help message?

File: bentoml/cli/deployment.py [hunk 8]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        )
+        if result.status.status_code != Status.OK:
+            _echo('Delete deployment {} failed'.format(name), CLI_COLOR_ERROR)
+            display_response_status_error(result.status)
+        else:
+            _echo('Successful delete deployment {}'.format(name), CLI_COLOR_SUCCESS)
+
+    @deploy.command()
+    @click.option("--name", type=click.STRING, help="Deployment's name", required=True)
+    @click.option("--ouput-format", type=click.STRING)
```
Comment (parano) [inline bentoml/cli/deployment.py]: remove not implemented options

File: bentoml/db.py [hunk 9]
```diff
@@ -27,7 +27,9 @@
 
 # sql alchemy config
 engine = create_engine(
-    config.get('db', 'engine'), echo=False, connect_args={'check_same_thread': False}
+    config.get('db', 'engine').strip("\'"),
```
Comment (parano) [inline bentoml/db.py]: discussed offline - remove this change

File: bentoml/config/default_bentoml.cfg [hunk 10]
```diff
@@ -19,6 +19,9 @@
 [core]
 usage_tracking = true
 
+[yatai]
+url = 'localhost:50051'
```
Comment (parano) [inline bentoml/config/default_bentoml.cfg]: remove

File: bentoml/cli/deployment.py [hunk 11]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+            _echo('Apply deployment {} failed'.format(bento_name), CLI_COLOR_ERROR)
+            display_response_status_error(result.status)
+        else:
+            _echo(
+                'Successful apply deployment {}'.format(bento_name), CLI_COLOR_SUCCESS
+            )
+            display_deployment_info(result.deployment)
+
+    @deploy.command()
+    @click.option("--name", type=click.STRING, help="Deployment's name", required=True)
```
Comment (parano) [inline bentoml/cli/deployment.py]: nit: `help="deployment name"`

File: bentoml/cli/deployment.py [hunk 12]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+                    namespace=namespace,
+                    name=deployment_name,
+                    annotations=parse_key_value_string(annotations),
+                    labels=parse_key_value_string(labels),
+                    spec=spec,
+                )
+            )
+        )
+        if result.status.status_code != Status.OK:
+            _echo('Apply deployment {} failed'.format(bento_name), CLI_COLOR_ERROR)
```
Comment (parano) [inline bentoml/cli/deployment.py]: s/bento_name/deployment_name

File: bentoml/cli/deployment.py [hunk 13]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+            display_response_status_error(result.status)
```
Comment (parano) [inline bentoml/cli/deployment.py]: Let's put all the error output into one line? for example:
```
Apply deployment ABC failed: "Error ALREADY_EXISTS: Deployment named ABC already exists"
```
note that the error status code doesn't mean anything to the user

File: bentoml/cli/deployment.py [hunk 14]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        else:
+            _echo(
+                'Successful apply deployment {}'.format(bento_name), CLI_COLOR_SUCCESS
```
Comment (parano) [inline bentoml/cli/deployment.py]: s/bento_name/deployment_name

File: bentoml/cli/deployment.py [hunk 15]
```diff
@@ -213,3 +227,225 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+    @click.option(
+        "--platform",
+        type=click.Choice(
+            ["aws_lambda", "gcp_function", "aws_sagemaker", "kubernetes", "custom"]
+        ),
+        required=True,
+        help="Target platform that Bento archive is going to deployed to",
+    )
+    @click.option("--namespace", type=click.STRING, help="Deployment's namespace")
+    @click.option("--labels", type=click.STRING)
```
Comment (parano) [inline bentoml/cli/deployment.py]: add help message, and examples? something like '--labels=foo=123,env=staging'

Comment (parano) [reply bentoml/cli/deployment.py]: make sure to denote that if an option is only effective for certain deploy platform

File: bentoml/cli/deployment.py (hunk 5 above)
Comment (yubozhao) [reply bentoml/cli/deployment.py]: Good suggestion

File: bentoml/cli/deployment.py (hunk 6 above)
Comment (yubozhao) [reply bentoml/cli/deployment.py]: good point, will put guard on that

File: bentoml/cli/deployment.py [hunk 16]
```diff
@@ -213,3 +227,240 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+def get_deployment_operator_type(platform):
+    return DeploymentOperator.Value(platform.upper())
+
+
+def display_response_status_error(status):
+    _echo('Error message: {}'.format(status.error_message), CLI_COLOR_ERROR)
+
+
+def display_deployment_info(deployment, output):
+    _echo('Deployment {} info\n\n'.format(deployment.name))
```
Comment (parano) [inline bentoml/cli/deployment.py]: remove this line? - most of time when people use json/yaml output, they may pipe the output to other CLI program

File: bentoml/cli/deployment.py [hunk 17]
```diff
@@ -213,3 +227,241 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
         deployment.check_status()
 
     return cli
+
+
+def parse_key_value_string(key_value_string):
```
Comment (parano) [inline bentoml/cli/deployment.py]:
```
def parse_key_value_pairs(key_value_pairs_str):
    result = {}
    if key_value_pairs_str:
        for key_value_pair in key_value_pairs_str.split(','):
              key, value = key_value_pair.split('=')
              key = key.strip()
              value = value.strip()
              if key in result:
                logger.warning('duplicated key "%s" found string map parameter')
              result[key] = value
      return result
```

File: bentoml/utils/__init__.py [hunk 18]
```diff
@@ -54,3 +58,22 @@ def isidentifier(s):
     except AttributeError:
         # str#isidentifier is only available in python 3
         return re.match(r"[A-Za-z_][A-Za-z_0-9]*\Z", s) is not None
+
+
+def pb_to_json(message, output_format='obj'):
+    result = MessageToJson(message)
+    if output_format == 'string':
+        return result
+    return json.loads(result)
```
Comment (parano) [inline bentoml/utils/__init__.py]: isn't this just MessageToDict ?

File: bentoml/utils/__init__.py [hunk 19]
```diff
@@ -54,3 +58,22 @@ def isidentifier(s):
...
+
+
+def pb_to_yaml(message, output_format='obj'):
```
Comment (parano) [inline bentoml/utils/__init__.py]: I would prefer to just remove output_format option here and simply return yaml string - if you need to modify output yaml, might be better just modify the proto message before converting to yaml

File: bentoml/cli/deployment.py [hunk 20]
```diff
@@ -213,3 +227,241 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        if result.status.status_code != Status.OK:
+            _echo('Get deployment {} failed'.format(name), CLI_COLOR_ERROR)
+            display_response_status_error(result.status)
+        else:
+            display_deployment_info(result.deployment, output)
+
+    @deploy.command()
+    @click.option("--name", type=click.STRING, help="Deployment name", required=True)
+    @click.option('--output', type=click.Choice(['json', 'yaml']), default='json')
+    def describe(name, output=None):
```
Comment (parano) [inline bentoml/cli/deployment.py]: nit - remove `=None` if you have `default='json`

File: bentoml/cli/deployment.py [hunk 21]
```diff
@@ -213,3 +231,255 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+def get_deployment_sub_command(cli):
+    @click.group()
+    def deploy():
+        pass
+
+    @deploy.command(
+        short_help="Apply deployment configuration",
+        context_settings=dict(ignore_unknown_options=True, allow_extra_args=True),
+    )
+    @click.argument('--bento-tag', type=click.STRING, required=True)
```
Comment (parano) [inline bentoml/cli/deployment.py]: add doc and example tag

File: bentoml/cli/deployment.py [hunk 22]
```diff
@@ -213,3 +231,255 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        stage=None,
+        instance_type=None,
+        instance_count=None,
+        api_name=None,
+        kube_namespace=None,
+        replicas=None,
+        service_name=None,
+        service_type=None,
+    ):
+        bento_name, bento_verison = parse_bento_tag(bento_tag)
```
Comment (parano) [inline bentoml/cli/deployment.py]: let's use a validation callback for bento_tag argument https://click.palletsprojects.com/en/5.x/options/#callbacks-for-validation

that way here can be simply `bento_name, bento_verison = bento_tag.split(':')`

File: bentoml/cli/deployment.py [hunk 23]
```diff
@@ -213,3 +231,255 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+    @click.option(
+        "--platform",
+        type=click.Choice(
+            ["aws_lambda", "gcp_function", "aws_sagemaker", "kubernetes", "custom"]
+        ),
+        required=True,
+        help="Target platform that Bento archive is going to deployed to",
+    )
+    @click.option("--namespace", type=click.STRING, help="Deployment's namespace")
+    @click.option("--labels", type=click.STRING, help="")
```
Comment (parano) [inline bentoml/cli/deployment.py]: help

File: bentoml/cli/deployment.py [hunk 24]
```diff
@@ -213,3 +231,255 @@ def check_deployment_status(archive_path, platform, region, stage, api_name):
...
+        return items[0], items[1]
+
+
+def get_deployment_sub_command(cli):
+    @click.group()
+    def deploy():
+        pass
+
+    @deploy.command(
+        short_help="Apply deployment configuration",
```
Comment (parano) [inline bentoml/cli/deployment.py]: nit: `Create or update a model serving deployment`

File: bentoml/cli/deployment.py (hunk 21 above)
Comment (parano) [inline bentoml/cli/deployment.py]: chatted offline, changing this to:
```bentoml deploy my-classifier --bento=my-classifier:1.03 --platform=aws_lambda```

**Pull request merged**